src/libraries/System.Text.Encodings.Web/src/System.Text.Encodings.Web.csproj

```diff
@@ -2,7 +2,6 @@
   <PropertyGroup>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <TargetFrameworks>$(NetCoreAppCurrent);$(NetCoreAppCurrent)-Browser;netcoreapp3.1;netstandard2.0;net461</TargetFrameworks>
```

Comment:
Looks like we are including browser here just for this little ifdef:
https://github.com/dotnet/runtime/blob/85fde0536b64c4901056b894f5184103f39cac63/src/libraries/System.Text.Encodings.Web/src/System/Text/Encodings/Web/OptimizedInboxTextEncoder.Ascii.cs#L25-L35

Doesn't seem worth it to me.  Also I don't like us having PNSE's buried in a library that has nothing to do with platform-specific functionality (Encoding).  Shouldn't we have some runtime feature flag that can detect if the platform supports this path and avoid even going down here?  @eerhardt @GrabYourPitchforks

Reply:
> Shouldn't we have some runtime feature flag that can detect

The issue is the `[StructLayout(LayoutKind.Explicit)]` attribute on this type. If the linker sees a used struct with that attribute, it can't trim any of the fields on it - since you are defining a `StructLayout`, you are probably using it in a P/Invoke, so trimming fields would break the P/Invoke.

Since the field can't be trimmed, it means that `Vector128<T>` can't be trimmed in `browser-wasm`.

Reply:
FYI - @vitek-karas @MichalStrehovsky @marek-safar. Looks related to https://github.com/mono/linker/pull/1309. There is some discussion there about this exact scenario.

Reply:
>  since you are defining a StructLayout, you are probably using it in a P/Invoke, so trimming fields would break the P/Invoke.

I believe @GrabYourPitchforks did this so that he could have a union structure where he could access the same bytes as a Vector as well.  Is there a way to annotate that it's safe to trim it?  Alternatively the linker could identify that this is a union, where it's safe to trim a field so long as the size of the struct remains the same.

Reply:
> I believe @GrabYourPitchforks did this so that he could have a union structure where he could access the same bytes as a Vector as well.

Can this use Unsafe.As or similar at the call site instead of explicit Vector128 inside the struct?

Reply:
> Can this use Unsafe.As or similar at the call site instead of explicit Vector128 inside the struct?

The presence of Vector128 changes the alignment requirement of the struct so dropping the field and using Unsafe.As would likely result in unaligned reads.

Looks like the problem was that the presence of Vector128 regresses the WASM size by ~4kB compressed. Vector128 is more fat than it needs to be. I assume it's caused by Equals/GetHashCode/ToString having too many dependencies.

Reply:
I opened https://github.com/dotnet/runtime/issues/53900 to discuss this in more detail.